File: java/src/ly/xstream/streaming/Connection.java
package ly.xstream.streaming;

import java.io.IOException;
import java.util.*;

import org.json.JSONException;
import org.json.JSONObject;

import com.clwillingham.socket.io.*;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;


class Connection implements MessageCallback, IRawMessageHandler{
	
	private ConnectionStates state;
	private HashMap<String,List<IRawMessageHandler>> bindings = new HashMap<String,List<IRawMessageHandler>>();
	private IOSocket socket;
	public String id;
	private List<String> securityTokens = new Vector<String>();
	private ILogger logger;
	private Integer subscriptionCount =0;
	private String appKey;
	
	public Connection(String appKey, String endPoint,ILogger logger){
		this.logger=logger;
		this.appKey = appKey;
		try{
			//NOTE this is not a secure connection
			//I couldn't get SSL and the Java NIO
			//to play together nicely
			logger.log("connecting to: "+endPoint);
			socket = new IOSocket(endPoint,this);
			setStatus(ConnectionStates.connecting);
			socket.connect();
		} catch(Exception ex){
			logger.handleError("problem connectiong to socket: ",ex);
			setStatus(ConnectionStates.failed);
			return;
		}
		
		bind("error",this);
	}
	
	public IClosable subscribe(final String channel, final Member member,final IDataUpdateHandler handler,final IMemberMessageHandler memberHandler){
		final String id = socket.getSessionID()+'|'+channel+'|'+subscriptionCount.toString();
		subscriptionCount++;
		final SubscriptionData subscriptionData =  new SubscriptionData(this.appKey,channel,id,member);
		
		
		onActive(new IConnectionActiveCallback() {
			@Override
			public void onConnectionActive(Connection connection) {
				try {
					connection.send("internal:subscribe",subscriptionData );
				} catch (Exception e) {
					logger.handleError("problem subscribing", e);
				}
			}
		}, true);
		
		final IRawMessageHandler rawHandler = new IRawMessageHandler() {
			@Override
			public void handleMessae(String eventName, JSONObject data) {
				DataMessages messages = StreamingClient.XStreaemlyGson.fromJson(data.toString(), DataMessages.class);
				handler.handleUpdate(messages);
			}
		};
		
		final IRawMessageHandler presenceHandler = new IRawMessageHandler() {
			@Override
			public void handleMessae(String eventName, JSONObject data) {
				try{
					MembersMessage members = StreamingClient.XStreaemlyGson.fromJson(data.toString(), MembersMessage.class);
					memberHandler.handleMessage(members);
				} catch(JsonSyntaxException e){
					logger.handleError("problem parsing: "+data.toString(), e);
				}
				
			}
		};
		
		bind("internal:data:"+id, rawHandler);
		bind("internal:member:"+id,presenceHandler);
		
		IClosable closable= new IClosable() {
			@Override
			public void close() {
				unbind("internal:data:"+id, rawHandler);
				unbind("internal:member:"+id,presenceHandler);
				onActive(new IConnectionActiveCallback() {
					
					@Override
					public void onConnectionActive(Connection connection) {
						try {
							connection.send("internal:unsubscribe", subscriptionData);
						} catch (Exception e) {
							logger.handleError("problem unsubscribing", e);
						}
					}
				}, true);
				
			}
		};
		
		return closable;
	}
	
	IClosable subscribeActiveChannels(final IActiveChannelCallback callback){
		final String id = socket.getSessionID()+"|activeChannels"+Double.toString(Math.random());
		final ListActiveChannelsData data = new ListActiveChannelsData(id,this.appKey);
		final IRawMessageHandler messageHandler = new IRawMessageHandler() {
			@Override
			public void handleMessae(String eventName, JSONObject data) {
				ChannelDataWrapper channelDataWrapper = StreamingClient.XStreaemlyGson.fromJson(data.toString(), ChannelDataWrapper.class);
				for(ChannelData channelData: channelDataWrapper.channels){
					callback.channelChanged(channelData);
				}
			}
		};
		
		bind("internal:channel:"+id, messageHandler);
		
		IClosable result = new IClosable() {
			@Override
			public void close() {
				unbind("internal:channel:"+id, messageHandler);
				onActive(new IConnectionActiveCallback() {
					@Override
					public void onConnectionActive(Connection connection) {
						try {
							connection.send("internal:channels:unsubscribe", data);
						} catch (Exception e) {
							logger.handleError("problem unsubscribing to active channels", e);
						}
					}
				}, true);
			}
		};
		
		onActive(new IConnectionActiveCallback() {
			@Override
			public void onConnectionActive(Connection connection) {
				try {
					connection.send("internal:channels:subscribe",data );
				} catch (Exception e) {
					logger.handleError("problem subscribing to channels", e);
				}
			}
		}, true);
		
		return result;
	}
	
	public void bind(String eventName,IRawMessageHandler handler){
		if(!bindings.containsKey(eventName)){
			bindings.put(eventName, new Vector<IRawMessageHandler>());
		}
		
		bindings.get(eventName).add(handler);
	}
	
	public void unbind(String eventName,IRawMessageHandler handler){
		if(bindings.containsKey(eventName)){
			bindings.get(eventName).remove(handler);
		}
	}
	
	public void onActive(final IConnectionActiveCallback action,Boolean repeat){
		Boolean fired = false;
		
		if(state == ConnectionStates.connected){
			action.onConnectionActive(this);
			fired = true;
		}
		
		if(!fired || repeat){
			OnActiveClosure handler = new OnActiveClosure(repeat,this,action);
			
			bind("connected",handler);
		}
	}
	
	public void stop(){
		if(null!=socket){
            socket.disconnect();
            socket= null;
        }
	}
	
	public void fireEvent(String eventName,JSONObject data){
		if(bindings.containsKey(eventName)){
			for(IRawMessageHandler handler: bindings.get(eventName)){
				handler.handleMessae(eventName,  data);
			}
		}
		
		if(bindings.containsKey(null)){
			for(IRawMessageHandler handler: bindings.get(null)){
				handler.handleMessae(eventName,  data);
			}
		}
	}
	
	public void applySession(String token){
		securityTokens.add(token);
		
		if(state == ConnectionStates.awaitingSecurityClearence
				||state == ConnectionStates.securityCheckFailed
				||state == ConnectionStates.connected){
			sendSecurityToken(token);
		}
	}
	
	public void send(String event,Object data) throws IOException, JSONException{
		send(event,data,null);
	}
	
	public void send(String event,Object data,AckCallback callback) throws IOException, JSONException{
		Gson gson = StreamingClient.XStreaemlyGson;
				
		String json = gson.toJson(data);
		JSONObject obj = new JSONObject(json);
		if(null==callback){
			socket.emit(event, obj);
		} else {
			socket.emit(event, obj,callback);
		}
	}
	
	private void sendSecurityToken(String token){
		logger.log("sending security token: "+token);
		
		try{
			socket.emit("internal:securityToken",token);
		} catch (Exception ex){
			logger.handleError("problem sending security token: ", ex);
			setStatus(ConnectionStates.failed);
		}
	}
	
	private void setStatus(ConnectionStates status){
		state = status;
		logger.log("connection status set to: "+status);
		fireEvent(status.toString(),null);
	}
	
	private void handelSecurtyCallback(JSONObject json)
	{
		Gson gson = new Gson();
		SecurityCallback data= gson.fromJson(json.toString(), SecurityCallback.class);
		
		logger.log("got security call back: "+data.error+ " valid: "+data.valid);
		if(data.valid && state!=ConnectionStates.connected){
			setStatus(ConnectionStates.connected);
		} else if(!data.valid && state ==ConnectionStates.awaitingSecurityClearence){
			setStatus(ConnectionStates.securityCheckFailed);
		}
	}
	
	
	@Override
	public void on(String event, JSONObject... data) {
		if("internal:securityTokenResult".equals(event)){
			handelSecurtyCallback(data[0]);
		} else {
			fireEvent(event,data[0]);
		}
	}

	@Override
	public void onMessage(String message) {
		logger.log("got unexpected onMessage "+message);
	}

	@Override
	public void onMessage(JSONObject json) {
		logger.log("got unexpected onMessage(json) "+json.toString());
	}

	@Override
	public void onConnect() {
		id = socket.getSessionID();
		setStatus(ConnectionStates.awaitingSecurityClearence);
		for(String token: securityTokens){
			sendSecurityToken(token);
		}
		
	}

	@Override
	public void onDisconnect() {
		setStatus(ConnectionStates.disconnected);
	}

	@Override
	public void onConnectFailure() {
		logger.log("connection failure");
		setStatus(ConnectionStates.failed);
	}
	
	private class SubscriptionData
	{
		public SubscriptionData(String appKey,String channel,String id,Member member)
		{
			this.appKey = appKey;
			this.channel = channel;
			this.id = id;
			this.member  = member;
		}
		
		@SuppressWarnings("unused")//used via JSON
		public String appKey;
		@SuppressWarnings("unused")//used via JSON
		public String channel;
		@SuppressWarnings("unused")//used via JSON
		public String id;
		@SuppressWarnings("unused")//used via JSON
		public Member member;
	}

	private class OnActiveClosure implements IRawMessageHandler{
		private Boolean beenFired = false;
		
		private Boolean repeat;
		private Connection connection;
		private IConnectionActiveCallback action;
		
		public OnActiveClosure(Boolean repeat,Connection connection,IConnectionActiveCallback action){
			this.repeat=repeat;
			this.connection = connection;
			this.action=action;
		}
		
		@Override
		public void handleMessae(String eventName,JSONObject data) {
			 if(!beenFired || repeat){
				 action.onConnectionActive(connection);
				 beenFired = true;
			 }
			
		}
	}
	
	private class ListActiveChannelsData
	{
		public ListActiveChannelsData(String id, String appKey)
		{
			this.id = id;
			this.appKey=appKey;
		}
		
		@SuppressWarnings("unused")//used via JSON
		public String id;
		
		@SuppressWarnings("unused")//used via JSON
		public String appKey;
	}
	
	private class SecurityCallback{
		public String error;
		public Boolean valid;
	}

	@Override
	public void handleMessae(String eventName, JSONObject data) {
		if(eventName.equals("error")){
			logger.log("Got error: "+ data.toString());
		}
	}
}

File: node.js/lib/connection.js
var io = require('socket.io-client-xstreamly');
var env = require('./env');

module.exports = Connection;

function Connection(appKey) {
    var bindings = [];
    var subscriptionCount = 0;
    this.state = "initialized";
    this.id=0;
    this.securityTokens = [];
    
    var me = this;
    
    this.bind = function(eventName, callback) {
        bindings.push({ name: eventName, func: callback });
    };
    
    //sets the address that socket.io should connect to
    //can only be casued once
    this.setAddress = function(address){
      if(!this.socket){
        try {
          if (null != io && undefined != io) {
            env.log('trying to connect to '+address);
            me.socket = io.connect(address, {
              "force new connection": true,
            });

            me.setStatus("connecting");
            
            me.socket.on('connecting',function(transport){
              numberOfConnections++;
              env.log('atempting to connect via: '+transport);

            });
            var numberOfConnections = 0;
            me.socket.on('reconnect',function(transport,numTries){
              env.log('atempting to reconect via: '+transport+ ' for the '+numTries+' times');
              if(numberOfConnections>10){
                env.log('stopping trying to connect');
                me.socket.disconnect();
                me.setStatus("error");
              }
            });
          }
          else {
            me.setStatus("unavailible");
          }
                 
          me.socket.on('connect', function() {
            me.id=me.socket.socket.sessionid;
              me.setStatus('awaiting-security-clearance');
              for(var key in me.securityTokens){
                var token = me.securityTokens[key];
                if(token){
                  sendSecurityToken(token);
                }
              }
          });
          
          
          me.socket.on('disconnect', function() {
              me.setStatus("disconnected");
          });
          
          me.socket.on('error', function(msg) {
            if(typeof(msg)=="string"){
                env.log("socket error: " + msg);
              }
              else{
                env.log("socket error: "+JSON.stringify(msg));
              }
          });
          
          me.socket.on('internal:securityTokenResult',function(item){
            env.log('got security call back: '+item.error+' valid: '+item.valid);
            if(item.valid && me.state!=='connected'){
              me.setStatus('connected');
            }
            else if (!item.valid && me.state ==='awaiting-security-clearance'){
              me.setStatus('security-check-failed');
            }
          });
          
        }
        catch (ex) {
            env.log("problem connecting to socket " + ex);
            this.setStatus("failed");
            return;
        }
      }
    };

    this.onActive = function(action, repeat) {
        var fired = false;
        if (this.state === "connected") {
            action(this);
            fired = true;
        }

        if (!fired || repeat) {
            me.bind("connected", function() {
                if (!fired || repeat) {
                    fired = true;
                    action(me);
                }
            });
        }
    }
    
    this.stop = function(){
    	if(me.socket){
            me.socket.disconnect();
        }
    }

    var sendSecurityToken = function (token){
  	  env.log('sending security token: '+token);
  	  me.socket.emit('internal:securityToken',token);
    }
    
    this.applySession = function(token){
    	me.securityTokens.push(token);
    	
    	//if we are already connected send the token over
    	if(me.state==='awaiting-security-clearance'
    	||me.state==='security-check-failed'
    	||me.state==='connected'){
    		sendSecurityToken(token);
    	}
    };

    this.setStatus = function(status) {
      env.log("changing status from " + this.state + " to " + status);
      this.state = status;
      fireEvent(status);
    };
    
    this.subscribe = function(name,includePersistedMessages,member,dataCallback,memberCallback){
      var me = this;
      subscriptionCount++;
      var id = me.id +'|'+name+'|'+subscriptionCount;
      
      me.socket.on('internal:data:'+id,dataCallback);
      me.socket.on('internal:member:'+id,memberCallback);
      
      var subscriptionData = {appKey:appKey,channel:name,id:id,member:member,includePersistedMessages:includePersistedMessages};
      
      me.socket.emit('internal:subscribe',subscriptionData);
      
      return {
        close: function(){
          me.socket.emit('internal:unsubscribe',subscriptionData);
        }
      };
    };

    function fireEvent(eventName, data) {
      env.log("firing event: " + eventName);
      for (var key in bindings) {
        if(bindings.hasOwnProperty(key)){
          var event = bindings[key].name;
          if (event == eventName) {
              bindings[key].func(data);
          }
          else if (event === undefined) {
              bindings[key].func(eventName, data);
          }
        }
      }
    }
    
    this.setStatus("waiting-for-address");
}
